sqlx-sqlite/src/connection/mod.rs

```diff
@@ -112,6 +115,50 @@ impl SqliteConnection {
 
         Ok(LockedSqliteHandle { guard })
     }
+
+    /// Serializes the SQLite database to an in-memory buffer.
+    ///
+    /// This function allows you to serialize an SQLite database to an in-memory buffer.
+    /// The output is the same bytes as if the database were to be written to disk.
+    pub async fn serialize_to_buffer(&mut self) -> Result<SerializedDatabase, Error> {
+        let locked = self.lock_handle().await?;
+
+        unsafe {
+            let mut size: sqlite3_int64 = 0;
+            let data_ptr = sqlite3_serialize(
+                locked.guard.handle.as_ptr(),
+                std::ptr::null(),
+                &mut size as *mut _,
+                0,
+            );
+
+            Ok(SerializedDatabase::new(data_ptr, size as usize))
+        }
+    }
+    
+    /// Deserializes a byte slice into a read-only SQLite database.
+    ///
+    /// This function allows you to take a byte slice representing a serialized database,
+    /// and load it as a read-only SQLite database.
+    pub async fn deserialize_to_readonly_db(&mut self, data: &[u8]) -> Result<(), Error> {
+        let locked = self.lock_handle().await?;
+
+        unsafe {
+            let status = sqlite3_deserialize(
```

Comment:
It's not explicitly stated but certain parts of the documentation of this function suggests that the buffer needs to stay around for SQLite to read into it:

> The serialized database P is N bytes in size. M is the size of the buffer P, which might be larger than N. If M is larger than N, and the SQLITE_DESERIALIZE_READONLY bit is not set in F, then SQLite is permitted to add content to the in-memory database as long as the total size does not exceed M bytes.
> 
> If the SQLITE_DESERIALIZE_FREEONCLOSE bit is set in F, then SQLite will invoke sqlite3_free() on the serialization buffer when the database connection closes. If the SQLITE_DESERIALIZE_RESIZEABLE bit is set, then SQLite will try to increase the buffer size using sqlite3_realloc64() if writes on the database cause it to grow larger than M bytes.

Looking at [the implementation of `sqlite3_deserialize()`](https://github.com/sqlite/sqlite/blob/8aaf63c6ac8b8292c0ecead0d2b04b68e9e6be78/src/memdb.c#L831C9-L831C9) seems to reinforce this. The buffer pointer `pData` is only referenced a few times, one of which is squirreling it away in a `MemStore` structure associated with the database connection, and the others have to do with the handling of `SQLITE_DESERIALIZE_FREEONCLOSE`.

Therefore this API is inherently unsound as the lifetime of `data` is not linked at all to the lifetime of the connection.

The only fix for this would be to take ownership of the buffer and keep it with the connection at all times, and ensure it's not freed before the connection.

Reply:
It would actually be pretty cool to be able to, say, set a `Arc<[u8]>` in `SqliteOptions` to deserialize from a read-only buffer so you could have a pool of connections all reading from the same buffer.